node/src/test/scala/coop/rchain/node/mergeablity/MergeabilityRules.scala

```diff
@@ -0,0 +1,297 @@
+    * as `errored`. Errored-deploys should make no changes to the state which should a mergeable case.
+    */
+  def InfiniteLoop(left: Rho*)(
+      right: Rho*
+  )(base: Rho*)(mergedState: State = State(Seq.empty, Seq.empty, Seq.empty)): Unit =
+    MergeableCase(left.reduce(_ | _))(right.reduce(_ | _))(base.reduce(_ | _))(mergedState)
+
+  private def getDataContinuationOnChannel0[F[_]: Sync](
+      runtime: RhoRuntime[F],
+      stateHash: Blake2b256Hash
+  ) = {
+    import coop.rchain.models.rholang.{implicits => toPar}
+    for {
+      _            <- runtime.reset(stateHash)
+      data         <- runtime.getData(toPar(Expr(GInt(0))))
+      continuation <- runtime.getContinuation(Seq(toPar(Expr(GInt(0)))))
+      join         <- runtime.getJoins(toPar(Expr(GInt(0))))
+    } yield State(data, continuation, join)
+  }
+
+  private[this] def checkConflictAndMergeState(
+      base: Rho,
+      left: Rho,
+      right: Rho,
+      isConflict: Boolean,
+      mergedStateResult: State,
+      rejectRight: Boolean
+  )(
+      implicit file: sourcecode.File,
+      line: sourcecode.Line
+  ): Task[Unit] = {
+
+    case class MergingNode(index: BlockIndex, isFinalized: Boolean, postState: Blake2b256Hash)
+
+    val baseDeploy = ConstructDeploy.sourceDeploy(base.value, 1L, phloLimit = 500)
+    val leftDeploy = ConstructDeploy.sourceDeploy(left.value, 2L, phloLimit = 500)
+    val rightDeploy =
+      ConstructDeploy.sourceDeploy(
+        right.value,
+        3L,
+        phloLimit = 500,
+        sec = ConstructDeploy.defaultSec2
+      )
+    implicit val concurrent                = Concurrent[Task]
+    implicit val metricsEff: Metrics[Task] = new Metrics.MetricsNOP[Task]
+    implicit val noopSpan: Span[Task]      = NoopSpan[Task]()
+    implicit val logger: Log[Task]         = Log.log[Task]
+    computeMergeCase[Task](
+      Seq(baseDeploy),
+      Seq(leftDeploy),
+      Seq(rightDeploy),
+      (runtime, _, mergedState) => {
+        val mergedRoot      = mergedState._1
+        val rejectedDeploys = mergedState._2
+        for {
+          dataContinuationAtMergedState <- getDataContinuationOnChannel0(
+                                            runtime,
+                                            mergedRoot
+                                          )
+          errMsg = s""" FAILED
+                  | base = ${baseDeploy}
+                  | left   = ${leftDeploy}
+                  | right   = ${rightDeploy}
+                  | Conflict: ${rejectedDeploys.nonEmpty} should be ${isConflict} with ${rejectedDeploys}
+                  | isRejectRight: ${rejectRight}
+                  | Merged state
+                  | ${dataContinuationAtMergedState}
+                  | should be
+                  | ${mergedStateResult}
+                  |
+                  | conflicts found: ${mergedState._2.size}
+                  |
+                  | go see it at ${file.value}:${line.value}
+                  | """.stripMargin
```

Comment:
Scala Test has support for custom `Matcher`s which can provide correct source reference where assertion is created so there is no need for manual manipulation of source info.

Maybe you can create a custom matcher for `State` similarly to this one for empty stream. It can provide the same info but with default Scala Test failed behavior.
https://github.com/rchain/rchain/blob/cba9de55821380fdf4994c05a2320cd78e229f9c/casper/src/test/scala/coop/rchain/casper/util/scalatest/Fs2StreamMatchers.scala#L17-L30